proxy/round_tripper/proxy_round_tripper.go

```diff
@@ -233,13 +236,39 @@ func (rt *roundTripper) backendRoundTrip(
 
 	rt.combinedReporter.CaptureRoutingRequest(endpoint)
 	tr := GetRoundTripper(endpoint, rt.roundTripperFactory)
-	res, err := tr.RoundTrip(request)
+	res, err := rt.timedRoundTrip(tr, request)
 
 	// decrement connection stats
 	iter.PostRequest(endpoint)
 	return res, err
 }
 
+func (rt *roundTripper) timedRoundTrip(tr http.RoundTripper, request *http.Request) (*http.Response, error) {
+
+	if rt.endpointTimeout <= 0 {
+		return tr.RoundTrip(request)
+	}
+	reqCtx, cancel := context.WithTimeout(request.Context(), rt.endpointTimeout)
+	request = request.WithContext(reqCtx)
+	errored := make(chan struct{}, 1)
+
+	go func() {
+		defer cancel()
+		select {
+		case <-reqCtx.Done():
+		case <-errored:
+		}
+	}()
+
+	resp, err := tr.RoundTrip(request)
+	if err != nil {
+		cancel()
+		errored <- struct{}{}
```

Comment:
Shouldn't we just be cancelling here and then returning there error (with an empty response). Function and its caller are already both setup that way.

Reply:
Sure we could return the resp, and err but it is the next line of code and we would have to return the same thing in all situations. The reason we need to check the error is to call cancel() and stop the goroutine. I do see that instead of calling cancel here we can expect the goroutine to do that instead so here we only send a signal on the error channel.